### PR TITLE
C++ API parity: at::Tensor::detach

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26568 C++ API parity: at::Tensor::register_hook
* #26332 C++ API parity: at::Tensor::requires_grad_
* **#26251 C++ API parity: at::Tensor::detach**

Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

### CODE REVIEW

File: test/cpp/api/tensor.cpp [hunk 1]
```diff
@@ -406,3 +406,31 @@ TEST(TensorTest, Requires_grad_) {
   ASSERT_THROWS_WITH(y.requires_grad_(false),
     "requires_grad_ is not implemented for Tensor");
 }
+
+TEST(TensorTest, Detach) {
+  auto x = torch::tensor({5}, at::TensorOptions().requires_grad(true));
+  auto y = x * x;
+  const auto y_detached = y.detach();
+  ASSERT_FALSE(y.is_leaf());
+  ASSERT_TRUE(y_detached.is_leaf());
```
Comment (yf225) [inline test/cpp/api/tensor.cpp]: We might want to check `y_detached.requires_grad() == false`, based on https://pytorch.org/docs/stable/tensors.html#torch.Tensor.detach `The result will never require gradient`.

File: test/cpp/api/tensor.cpp [hunk 2]
```diff
@@ -406,3 +406,31 @@ TEST(TensorTest, Requires_grad_) {
...
+
+  x = at::tensor({5}, at::TensorOptions().requires_grad(false));
+  y = x * x;
+  const auto message = "detach is not implemented for Tensor";
+  ASSERT_THROWS_WITH(x.detach(), message);
+  ASSERT_THROWS_WITH(y.detach(), message);
+}
+
+TEST(TensorTest, Detach_) {
```
Comment (yf225) [inline test/cpp/api/tensor.cpp]: I feel that we might want to name this `DetachInplace`, to preserve the naming convention (of not using underscore). Same for other tests as well.

**Review by yf225 — APPROVED** [review]
Thanks a lot for the thorough tests @pbelevich !
